### PR TITLE
Model rendering

### PR DESCRIPTION
**Pull request by zhang-jm**
#2 

- Load shaders
- Load obj model
- Render obj model

### CODE REVIEW

**Review by julyzamora — APPROVED** [review]

File: src/graphics/src/model.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    std::cout << "Loading model at path: " << path << std::endl;
+
+    // read file via ASSIMP
+    Assimp::Importer importer;
+    const aiScene* const scene = importer.ReadFile(path, aiProcess_Triangulate);
+
+    // check for errors
+    if (!scene || scene->mFlags == AI_SCENE_FLAGS_INCOMPLETE || !scene->mRootNode)
+    {
+        std::cout << "Error: Assimp could not load model -  " << importer.GetErrorString() << std::endl;
```
Comment (julyzamora) [inline src/graphics/src/model.cpp:33]: Maybe we could add some perror here and future print statements to catch errors. I guess terror is implemented with built in knowledge meaning of errno lawl?

File: src/graphics/src/shader_loader.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    // save shader if successful load
+    int programID = this->ParseShader(s);
+    if (programID >= 0) {
+        shaderCache.emplace(s, programID);
+    }
+
+    return programID;
+}
+
+int ShaderLoader::LoadShader(const char * vertexShaderPath, const char * fragmentShaderPath, const char * geometryShaderPath) {
```
Comment (julyzamora) [inline src/graphics/src/shader_loader.cpp]: V long line lawl

File: src/graphics/src/model.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                std::cout << "diffuse" << std::endl;
+                vertex.color = glm::vec4(diffuse.r, diffuse.g, diffuse.b, diffuse.a);
+            }
+        }
+
+        vertices.push_back(vertex);
+    }
+
+    // get indices for vertexes
+    for (GLuint i = 0; i < mesh->mNumFaces; i++)
```
Comment (julyzamora) [inline src/graphics/src/model.cpp:103]: Consider separating for loops by creating another vector to hold the faces of the mesh adds space complexity O(mNumFaces) but time will be O(m + n). Not worth if not called that much though. Unsure about this though maybe we can discuss if it's worth trying to optimize?

File: src/main_loops/src/client_loop.cpp [hunk 4]
```diff
@@ -53,6 +65,13 @@ int main(int argc, char * argv[]) {
 	glfwSetKeyCallback(mWindow, key_callback);
 	glfwSetMouseButtonCallback(mWindow, mouse_click_callback);
 
+    ShaderLoader loader;
+
+    projection = glm::perspective(45.0f, (float)window_width / (float)window_height, 0.1f, 1000.0f);
+    view = glm::lookAt(cam_pos, cam_look_at, cam_up);
+
+    Model m(loader, "../models/cube/cube.obj", "../src/graphics/shaders/playershader.vert", "../src/graphics/shaders/playershader.frag");
```
Comment (julyzamora) [inline src/main_loops/src/client_loop.cpp:73]: Maybe we can have a file to declare paths or make configurable. As long as the paths are defined somewhere. Would help if we have a lot of models

Comment (zhang-jm) [reply src/main_loops/src/client_loop.cpp:73]: I agree, I think this is something that can be done with #5 ?

**Pull request merged**